File: 1-D DP/Longest Increasing Subsequence/Standard LIS/solution.py
from typing import List
import bisect

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        tails = []
        for x in nums:
            idx = bisect.bisect_left(tails, x)
            if idx < len(tails):
                tails[idx] = x
            else:
                tails.append(x)
        return len(tails)

if __name__ == "__main__":
    sol = Solution()
    
    print(f"Test Case 1: {sol.lengthOfLIS([10,9,2,5,3,7,101,18])}") # 4
    print(f"Test Case 2: {sol.lengthOfLIS([0,1,0,3,2,3])}") # 4
    print(f"Test Case 3: {sol.lengthOfLIS([7,7,7,7,7,7,7])}") # 1

File: 1-D DP/Longest Increasing Subsequence/Standard LIS/solution.java
import java.util.ArrayList;
import java.util.List;
import java.util.Collections;

class Solution {
    public int lengthOfLIS(int[] nums) {
        if (nums.length == 0) return 0;
        
        List<Integer> tails = new ArrayList<>();
        
        for (int num : nums) {
            int idx = Collections.binarySearch(tails, num);
            if (idx < 0) idx = -(idx + 1); 
            
            if (idx == tails.size()) {
                tails.add(num);
            } else {
                tails.set(idx, num);
            }
        }
        
        return tails.size();
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        
        int[] nums1 = {10,9,2,5,3,7,101,18};
        System.out.println("Test Case 1: " + sol.lengthOfLIS(nums1)); // 4
        
        int[] nums2 = {0,1,0,3,2,3};
        System.out.println("Test Case 2: " + sol.lengthOfLIS(nums2)); // 4
        
        int[] nums3 = {7,7,7,7,7,7,7};
        System.out.println("Test Case 3: " + sol.lengthOfLIS(nums3)); // 1
    }
}
